iree/compiler/Dialect/Stream/Transforms/ElideAsyncCopies.cpp

```diff
@@ -0,0 +1,531 @@
+
+  // Not safe.
+  LLVM_DEBUG(llvm::dbgs() << "  - clone source cannot be elided\n");
+  return false;
+}
+
+// Tries to elide |cloneOp| by replacing all uses with its source if safe.
+// Returns true if the op was elided.
+static bool tryElideCloneOp(IREE::Stream::AsyncCloneOp cloneOp,
+                            LastUseAnalysis &analysis) {
+  if (!isSafeToElideCloneOp(cloneOp, analysis)) return false;
+  cloneOp.replaceAllUsesWith(cloneOp.source());
+  cloneOp.erase();
+  return true;
+}
+
+// Tries to elide copies nested within |region| when safe.
+// Returns true if any ops were elided.
+static bool tryElideAsyncCopiesInRegion(Region &region,
+                                        LastUseAnalysis &analysis) {
+  bool didChange = false;
+  for (auto &block : region) {
+    for (auto cloneOp : llvm::make_early_inc_range(
+             block.getOps<IREE::Stream::AsyncCloneOp>())) {
+      if (!isSafeToElideCloneOp(cloneOp, analysis)) continue;
+      cloneOp.replaceAllUsesWith(cloneOp.source());
+      cloneOp.erase();
+      didChange = true;
+    }
+  }
+  return didChange;
+}
+
+//===----------------------------------------------------------------------===//
+// -iree-stream-elide-async-copies
+//===----------------------------------------------------------------------===//
+
+// Elides async copies that perform no meaningful work - such as clones of the
+// last use of a value. This is designed to be run after
+// -iree-stream-materialize-copy-on-write to clean up the copies it introduces
+// but will also pick up any copies that came from the frontend.
+//
+// This should never remove copies that are required for correctness: we err on
+// the side of leaving copies when we cannot perform full analysis.
+//
+// This operates using a whole-program data flow analysis to first determine
+// which block arguments have move semantics (they are passed the last use of
+// a resource) and the last users of all cloned values. Once analyzed all copies
+// in the program are checked to see if they can be safely removed and if so are
+// rerouted to the cloned source value. This process repeats until no more
+// copies are elided: we are guaranteed to reach a fixed point as we are only
+// removing copies in this pass and not introducing any new ops.
+class ElideAsyncCopiesPass : public ElideAsyncCopiesBase<ElideAsyncCopiesPass> {
+ public:
+  ElideAsyncCopiesPass() = default;
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<IREE::Stream::StreamDialect>();
+    registry.insert<IREE::Util::UtilDialect>();
+  }
+
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    if (moduleOp.getBody()->empty()) return;
+
+    // Try analyzing the program and eliding the unneeded copies until we reach
+    // a fixed point (no more copies can be elided).
+    unsigned maxIterationCount = 30;
+    unsigned iterationCount = 0;
+    for (; iterationCount < maxIterationCount; ++iterationCount) {
```

Comment:
Note for future compiler optimization / debugging: may help to report the iteration count for these sorts of passes to statistics or Tracy.

I see the note below though :P

>    // this to happen without remarking. For now all our programs coverge in
      // just one or two iterations and this needs to be tuned with more complex
      // control flow.

Reply:
agreed! pass statistics would be good for this